// topnav/topnav.js
const graphic = require('graphic')
const IO = require('io')
const statedb = require('STATE')
const default_data = require('./data.json')
/******************************************************************************
  OUR CONTRIBUTORS COMPONENT
******************************************************************************/
// ----------------------------------------
const shopts = { mode: 'closed' }
// ----------------------------------------
module.exports = topnav

async function topnav (opts) {
	// ----------------------------------------
	// ID + JSON STATE
	// ----------------------------------------
	const name = 'topnav'
	const status = {}
	const on = {
		inject,
		inject_all,
		scroll
	}

	const sdb = statedb()
	let data = sdb.get(opts.sid)
	if(!data){
    const {id} = sdb.add(default_data, opts.hub)
    data = {...default_data, id}
  }
	const {send, css_id} = await IO({id: data.id, name, type: 'comp', comp: name, hub: opts.hub, css: data.css}, on)
	// ----------------------------------------
	// OPTS
	// ----------------------------------------

	const playLogo = await graphic('playLogo', './src/node_modules/assets/svg/logo.svg')
	// ----------------------------------------
	// TEMPLATE
	// ----------------------------------------
	const el = document.createElement('div')
	const shadow = el.attachShadow(shopts)
	shadow.innerHTML = `
		<section class='topnav'>
				<a href="#top">${playLogo.outerHTML}</a>
				<nav class='menu'>
				</nav>
		</section>
	`
	// ----------------------------------------
	const menu = shadow.querySelector('.menu')
	const body = shadow.querySelector('section')
	menu.append(...data.links.map(make_link))
	const scrollUp = 'scrollUp'
	const scrollDown = 'scrollDown'
	let lastScroll = 0
	
	window.addEventListener('scroll', ()=> {
		if (window.innerWidth >= 1024) {
			let currentScroll = window.pageYOffset
			if (currentScroll < 1) {
					body.classList.remove(scrollUp)
					body.classList.remove(scrollDown)
					return
			}
			if (currentScroll > lastScroll && !body.classList.contains(scrollDown)) {
					body.classList.add(scrollDown)
					body.classList.remove(scrollUp)
			} else if (currentScroll < lastScroll) {
					body.classList.add(scrollUp)
					body.classList.remove(scrollDown)
			}
			lastScroll = currentScroll
		}
	})

	window.addEventListener('resize', ()=> {
		if (window.innerWidth <= 1024) {
			body.classList.remove(scrollUp)
			body.classList.remove(scrollDown)
		}
	})
	init_css()
	return el

	function click(url) {
		send({to:'index', type: 'jump', data: url })
	}
	async function inject ({ data }) {
		sheet.replaceSync(data)
		shadow.adoptedStyleSheets = [sheet]
	}
	function make_link(link){
		const a = document.createElement('a')
		a.href = `#${link.url}`
		a.textContent = link.text
		a.onclick = () => click(link.url)
		return a
	}
	async function init_css () {
		const pref = JSON.parse(localStorage.pref)
		const pref_shared = pref[name] || data.shared || [{ id: name }]
		const pref_uniq = pref[css_id] || data.uniq || []
		pref_shared.forEach(async v => inject_all({ data: await get_theme(v)}))
		pref_uniq.forEach(async v => inject({ data: await get_theme(v)}))
	}
	async function scroll () {
		el.scrollIntoView({behavior: 'smooth'})
		el.tabIndex = '0'
		el.focus()
		el.onblur = () => {
			el.tabIndex = '-1'
			el.onblur = null
		}
	}
	async function inject_all ({ data }) {
		const sheet = new CSSStyleSheet
		sheet.replaceSync(data)
		shadow.adoptedStyleSheets.push(sheet)
	}
	async function inject ({ data }){
		const style = document.createElement('style')
		style.innerHTML = data
		shadow.append(style)
	}
	async function get_theme ({local = true, theme = 'default', id}) {
		let theme_css
		if(local)
			theme_css = await (await fetch(`./src/node_modules/css/${theme}/${id}.css`)).text()
		else
			theme_css = JSON.parse(localStorage[theme])[id]
		return theme_css
	}
}
